### PR TITLE
move-package: add name resolution hook

### PR DESCRIPTION
**Pull request by kklas**
## Description

Adds a `custom_resolve_pkg_name` hook in order to allow for the dependency graph to discriminate between packages based on a dynamically resolved symbol (returned by the hook) rather than just package name. When the hook isn't defined it returns the package manifest name when so this doesn't affect Sui package resolution currently. This will make it possible to consolidate source and on-chain packages (which don't have a name) within a single graph.

This is part of the work to enable compiling against on-chain dependencies https://github.com/MystenLabs/sui/pull/14178.

## Test Plan 

Added unit tests.

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

**Comment by kklas** [comment]
@amnn @rvantonder I'm debating whether the `custom_resolve_pkg_name` (and corresponding variables) should be renamed to `custom_resolve_pkg_identifier` because "pkg_name" variables in dependency graph conflating between package name as defined in manifest and resolved name gets super confusing with later changes.

**Comment by rvantonder** [comment]
> conflating between package name as defined in manifest and resolved name gets super confusing with later changes.

A better name sounds good to me! Thoughts on `custom_resolve_pkg_manifest_name` or `custom_resolve_pkg_name_from_manifest`? I think more verbosity here will help, `identifier` comes across as something more general than the "real" description that causes confusion :-)

**Comment by rvantonder** [comment]
(kicking off CI)

**Comment by kklas** [comment]
Well the problem is really with referring to the thing as "package name" since with the resolution hook it's not necessarily a name anymore. E.g. with Sui on-chain resolution hooks it will be its original id.

So the first point of confusion is that dependency graph code uses types such as `PM::PackageName` in maps as package keys and the other is that variable names such as `pkg_name` are used when with the resolution hook it's not necessarilly a package name. This becomes especially tricky around places where the resolution hook gets called and propagated because all variable names become confusing when reading the code.

I was thinking we need something more generic than package name.

**Comment by rvantonder** [comment]
Got it! If this is a case of just renaming, I'm generally in favor (indicating `identifer` is good, with appropriate comments to explain this could refer to a source package name, versus an on-chain ID).

@kklas is your sense that there's a direction that this is more than renaming, i.e., we may want types to represent these? 

Either way, I think we're good to follow up with a renaming / refactor PR to be hygienic about this, and don't need to bundle it with this PR. @amnn to weigh in when he's back from PTO later this week, but I don't see this as needing to block this PR. I'll take a deeper look soon!

**Comment by kklas** [comment]
I think it would just be a renaming in dependency_graph and resolution_graph files. `PM::PackageName` is a type alias to `Symbol` so we can just introduce a new alias like `PackageIdentifier`. There's a lot of `pkg_name` variables in these files though (~130).

File: crates/sui-move-build/src/lib.rs
Comment (amnn) [inline crates/sui-move-build/src/lib.rs:605]: nit: replace qualified use with an import?

File: external-crates/move/crates/move-package/resolve_pkg_name_nested_dep_conflict/Move.resolved
Comment (amnn) [inline external-crates/move/crates/move-package/resolve_pkg_name_nested_dep_conflict/Move.resolved:5]: In this test, the address substitutions don't match, but also the names don't match -- are we expecting the names to match? (It seems like we might be testing something slightly different if they don't).

File: external-crates/move/crates/move-package/resolve_pkg_name_nested_dep_conflict/deps_only/C/Move.toml
Comment (amnn) [inline external-crates/move/crates/move-package/resolve_pkg_name_nested_dep_conflict/deps_only/C/Move.toml:5]: (I think the problem is that the `ADep` name is coming from the external resolver which doesn't provide an `ADep-renamed`, only an `ADep`).

File: external-crates/move/crates/move-package/resolve_pkg_name_override/Move.toml
Comment (amnn) [inline external-crates/move/crates/move-package/resolve_pkg_name_override/Move.toml:22]: nit: this will compile locally, but typically you don't want to override in this way (to a lower version), as it won't publish.  Not an issue for this test, but may become an issue later if we become cleverer about detecting bad versioning arrangements locally.

File: external-crates/move/crates/move-package/src/resolution/resolution_graph.rs
Comment (amnn) [inline external-crates/move/crates/move-package/src/resolution/resolution_graph.rs:119]: The semantics of this check have no changed slightly -- previously it was ensuring that there was a consistency between nodes in the graph data structure and in the package table, now it's doing that consistency check by looking at the dependencies in the package table which is slightly different.  Could you go into the details of why this change was necessary? (It doesn't seem wrong to me, it's just a little less direct, so it would help me to understand).

File: external-crates/move/crates/move-package/src/resolution/resolution_graph.rs
Comment (amnn) [inline external-crates/move/crates/move-package/src/resolution/resolution_graph.rs:346]: Is it worth encapsulating the `name` field and only exposing a function that will do the custom resolution?

**Review by amnn — APPROVED** [review]
Generally looks good to me, and I'm also supportive of the choice to generalise the name to `identifier` rather than `name`!

File: external-crates/move/crates/move-package/resolve_pkg_name_nested_dep_conflict/Move.resolved
Comment (kklas) [reply external-crates/move/crates/move-package/resolve_pkg_name_nested_dep_conflict/Move.resolved]: This one is based on [diamond_problem_dep_nested_dep_conflict](https://github.com/MystenLabs/sui/blob/308dad00c432ef76d6fe3d9183acb65e76c88433/external-crates/move/crates/move-package/tests/test_sources/diamond_problem_dep_nested_dep_conflict/Move.resolved) and is intended to check for address substitution (but actually checks both and will fail with the same message).

But good catch! The reason names are different is because package `A` with dependency `ADep` is resolved externally in the other branch, which means its subgraph is fully reconstructed from the lockfile so name resolution hook doesn't kick in. This actually highlights the issue (2) I described in https://github.com/MystenLabs/sui/pull/14178 and why we will need to force lockfile regeneration when on-chain hooks get enabled.

I've improved this test by adding a separate external resolver script for this https://github.com/MystenLabs/sui/pull/15315/commits/9c9781a08e419ef6c6dc11c1c251d7d52cc91f2a. 

Also while looking into this I've noticed a separate problem -- the `deps_equal` function which checks all dependencies of two packages match didn't respect the custom `PartialEq` implementation for `Package` https://github.com/MystenLabs/sui/blob/5e2e4985ebf49e5e3723582d4035e29bb4eb66ef/external-crates/move/crates/move-package/src/resolution/dependency_graph.rs#L124-L132 and the above test would fail even with correct address substitution because the resolver is set on one and not the other. I've fixed this with https://github.com/MystenLabs/sui/pull/15315/commits/5e2e4985ebf49e5e3723582d4035e29bb4eb66ef

**Comment by kklas** [comment]
I will change name to `identifier` at the end because otherwise there will be a lot of merge conflicts with commits that are ready from before.

File: external-crates/move/crates/move-package/src/resolution/resolution_graph.rs
Comment (amnn) [reply external-crates/move/crates/move-package/src/resolution/resolution_graph.rs:346]: Will leave it up to you!  It might still be worth creating an abstraction for the name in its use as a key, and then add a separate abstraction to expose the user-friendly name.

File: external-crates/move/crates/move-package/src/resolution/resolution_graph.rs
Comment (amnn) [reply external-crates/move/crates/move-package/src/resolution/resolution_graph.rs:119]: Ah that makes sense, thanks for clarifying. To say it back to make sure I've understood:

The previous implementation of the check confirmed that the keys in the package table matched the names in manifests, in order to check that the names given to dependencies matched the names packages gave themselves.  Now that the names in the package table are derived from the manifests, via the name resolution hook, the old implementation would be performing a redundant check so we have to literally check that the dependencies of all packages point back to a valid package in the table.

Comment (kklas) [reply external-crates/move/crates/move-package/src/resolution/resolution_graph.rs:119]: >  the old implementation would be performing a redundant check so we have to literally check that the dependencies of all packages point back to a valid package in the table.

The old implementation is performing package name from table (which is loaded from dependency section of parent) to package name from manifest equality check. As it is this wouldn't work with resolution hook because package name from table (which is now a resolved name) is different from one in manifest. And if you do a package name from table (resolved name) to `custom_resolve_pkg_name(&self.source_package)` check this is indeed a redundand check.

So what we need is dependency name from depency section in parent manfiest to name in manifest check for sanity, otherwise someone could have a `DeepBook12313 = { ... }` for the `DeepBook` package and everything would work (since these names are now irrelevant for dependency resolution unless your custom hook relies on that).

File: external-crates/move/crates/move-package/src/resolution/resolution_graph.rs
Comment (kklas) [reply external-crates/move/crates/move-package/src/resolution/resolution_graph.rs:346]: Let's leave this for last, I'll take a look at this when I do the name -> identifier renaming.

**Comment by rvantonder** [comment]
(kicking off CI)

**Comment by kklas** [comment]
@rvantonder what's the status?

**Pull request merged**
